src/main/java/com/iota/iri/network/NeighborRouterImpl.java

```diff
@@ -0,0 +1,895 @@
+package com.iota.iri.network;
+
+import com.iota.iri.conf.BaseIotaConfig;
+import com.iota.iri.conf.NetworkConfig;
+import com.iota.iri.conf.ProtocolConfig;
+import com.iota.iri.controllers.TransactionViewModel;
+import com.iota.iri.model.Hash;
+import com.iota.iri.network.neighbor.Neighbor;
+import com.iota.iri.network.neighbor.NeighborState;
+import com.iota.iri.network.neighbor.impl.NeighborImpl;
+import com.iota.iri.network.pipeline.TransactionProcessingPipeline;
+import com.iota.iri.network.pipeline.TransactionProcessingPipelineImpl;
+import com.iota.iri.network.protocol.Handshake;
+import com.iota.iri.network.protocol.Protocol;
+import com.iota.iri.utils.Converter;
+
+import java.io.IOException;
+import java.net.*;
+import java.nio.ByteBuffer;
+import java.nio.channels.*;
+import java.security.SecureRandom;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import net.openhft.hashing.LongHashFunction;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A NeighborRouter takes care of managing connections to {@link Neighbor} instances, executing reads and writes from/to
+ * neighbors and ensuring that wanted neighbors are connected. <br/>
+ * A neighbor is identified by its identity which is made up of the IP address and the neighbor's own server socket port
+ * for new incoming connections; for example: 153.59.34.101:15600. <br/>
+ * The NeighborRouter and foreign neighbor will first exchange their server socket port via a handshaking packet, in
+ * order to fully build up the identity between each other. If handshaking fails, the connection is dropped.
+ */
+public class NeighborRouterImpl implements NeighborRouter {
```

Comment:
Nothing has changed except the name? Right?

Reply:
Interface extraction was necessary so that guice was able to handle the cyclic dependency (as discussed). There were several more changes so that I could extract the interface:

* parseURI -> from `public static` to `private`
* isURIValid ->  from `public static` to `private`
* `public static long getTxCacheDigest` -> moved into `TransactionCacheDigester`
* Javadoc moved into interface

Reply:
Sounds good

I usually like to leave the `static` on private methods that don't use the class fields. Just because it makes it explicit that the method doesn't depend on the state of the object.

There is also the superstitious belief that `static` will  be ran faster (I think the JVM knows to optimize private methods well enough)

But can you return the `static` just because of the first reason?

Reply:
done